pygbm/gradient_boosting.py

```diff
@@ -293,25 +291,29 @@ def _should_stop(scores):
                 # tree.
                 return False
 
+            # The higher the tol, the more likely we are to early stop. We
+            # thus want to (artificially) increase the value of prev_score by
+            # tol percent and check if it's higher than the current score, in
+            # which case we early stop. As in scikit-learn scorers "higher is
+            # always better", the 'error' metrics such as `neg_mse` or
+            # `neg_log_loss` output negative values so the scores are
+            # negative. We thus need to multiply tol by -1 in those cases,
+            # else we would actually decrease prev_score instead of increasing
+            # it.
+
             current_score = scores[-1]  # score at current iteration
             previous_scores = scores[-self.n_iter_no_change:-1]
+            sign = self.scorer_._sign
             return all(
-                current_score < prev_score * (1 + self.tol * scorer._sign)
+                current_score < prev_score * (1 + self.tol * sign)
```

Comment:
Actually, the previous code is not the stopping criterion I want. May I suggest the following:


```python
reference_position = self.n_iter_no_change + 1
if len(scores) < reference_position:
    reference_score = scores[-reference_position]
    recent_scores = scores[-reference_position + 1:]
    recent_improvements = [score > reference_score * (1 - tol)
                           for score in recent_scores]
    return not any(recent_improvements)
```

Reply:
Thinking about it, I think we should use an absolute value for `tol`: it's not intuitive interpret it as a fraction of the reference score value as it will evolve over time.

Therefore I suggest the following instead:

```python
reference_position = self.n_iter_no_change + 1
if len(scores) < reference_position:
    reference_score = scores[-reference_position] - tol  # higher is always better
    recent_scores = scores[-reference_position + 1:]
    recent_improvements = [score > reference_score
                           for score in recent_scores]
    return not any(recent_improvements)
```

Reply:
Ok I'll use an absolute tol, I think that's what sklearn is doing as well, and we don't have to worry about the sign anymore.

------

Let me know if I made a mistake in the previous version but I think that 

```python
all(current_score < prev_score * (1 + tol) for prev_score in previous_scores)
```

is incorrect when `scorer._sign` is `-1`, because in this case `prev_score` is negative (the closer it is to 0, the better). Multiplying it by `(1 + tol)` will decrease it instead of increasing it. Hence the use of  `1 + tol * sign`.

Reply:
Did you mean

```python
reference_score = scores[-reference_position] + tol
```

with a plus?

My understanding is taht the higher the tol, the more likely we are to early stop.

Reply:
> The best I have so far is: ...

I like this plain English explanation.

> My understanding is that before (and in sklearn), the higher the tol, the more likely we are to early stop. 

Yes I agree. My code is wrong. This is indeed very hard to reason correctly about this thing. Here is my "final" suggestion:


```python
reference_position = self.n_iter_no_change + 1
if len(scores) < reference_position:
    # Remarks: a higher score is always better
    # Higher tol means that it will be hard for subsequent iteration to be
    # considered an improvement upon the reference score and therefore
    # more likely to early stop because of the lack of significant improvement.
    reference_score = scores[-reference_position] + tol
    recent_scores = scores[-reference_position + 1:]
    recent_improvements = [score > reference_score
                           for score in recent_scores]
    return not any(recent_improvements)
```

Reply:
> Personally I found the previous strategy easier to explain and to understand.

I cannot understand the logic of the previous strategy. I prefer my latest suggestion.

Reply:
Could you please extract `_should_stop(scores, n_iter_no_change, tol)` as a standalone function (or method) so that we can write unittest for it?


```python
def test_shoud_stop():
    # not enough iterations
    assert not _should_stop([], n_iter_no_change=1, tol=0.001)

    assert not _should_stop([1, 1, 1], n_iter_no_change=5, tol=0.001)
    assert not _should_stop([1] * 5, n_iter_no_change=5, tol=0.001)

    # still making significant progress up to tol
    assert not _should_stop([1, 2, 3, 4, 5, 6], n_iter_no_change=5, tol=0.001)
    assert not _should_stop([1, 2, 3, 4, 5, 6], n_iter_no_change=5, tol=0.)
    assert not _should_stop([1, 2, 3, 4, 5, 6], n_iter_no_change=5, tol=0.999)

    # no significant progress according to tol
    assert _should_stop([1] * 6, n_iter_no_change=5, tol=0.)
    assert _should_stop([1] * 6, n_iter_no_change=5, tol=0.001)
    assert _should_stop([1, 2, 3, 4, 5, 6], n_iter_no_change=5, tol=2)

```